# RNAseq differential expression R notebook 

This R notebook can be used for differential expression and functional enrichment analyses. It was developed for the [2023 Introduction to RNAseq workshop](https://sydney-informatics-hub.github.io/rnaseq-workshop-2023/) (11-12th October). It accepts a merged gene count matrix generated by [Salmon/1.10.1](https://combine-lab.github.io/salmon/). 

This notebook requires a number of packages. You can execute it with our [RStudio for RNAseq container](https://github.com/Sydney-Informatics-Hub/Rstudio-rnaseq-contained) which has all required packages preinstalled. For more information on this notebook, please see the [training materials for our Introduction to RNAseq workshop](https://sydney-informatics-hub.github.io/rnaseq-workshop-2023/). A copy of the count matrix used in the workshop and rendered HTML file are also included in this repository in the `data/` directory.  

# Quickstart

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/#https://github.com/Sydney-Informatics-Hub/rna-differential-expression-Rnotebook)

# Acknowledgements 

The Sydney Informatics Hub is an official node of the Australian BioCommons, and has worked in partnership with National Computational Infrastructure, Pawsey Supercomputing Research Centre, and AARNet to create command-line resources that make bioinformatics more accessible for life scientists. 

Acknowledgements (and co-authorship, where appropriate) are an important way for us to demonstrate the value we bring to your research. Your research outcomes are vital for ongoing funding of the Sydney Informatics Hub and national compute facilities. We suggest including the following acknowledgement in any publications that follow from this work:

"The authors acknowledge the technical assistance provided by the Sydney Informatics Hub, a Core Research Facility of the University of Sydney and the Australian BioCommons which is enabled by NCRIS via Bioplatforms Australia".
