# Rmarkdown for nfcore-rnaseq 1023
This Rmarkdown file was developed for Day-2 of the workshop and perform analysis to convert the raw counts to Differentially Expressed Genes and Functional Enrichments 

## Quickstart
1. **RStudio version**:  `4.1.0`
1. Clone the repository
```
cd /PROJECT_DIR
git clone https://github.com/Sydney-Informatics-Hub/rnaseq_DE_Day2_1023.git`
```
2. Open the file `rnaseq_DE_analysis_Day2.Rmd` in RStudio and excecute it using the RStudio environment.
3. The script uses multiple R-libraries. A RStudio singularity container with pre-installed R-libraries is available at [dockerhub registry](https://hub.docker.com/repository/docker/sydneyinformaticshub/rnaseq-rstudio/general). Instructions to use this singularity container on Pawsey-Nimbus VM instance can be found [here](https://github.com/Sydney-Informatics-Hub/spaceranger-contained) 
4. Alternatively the R-libraries can be installed locally.    
